### PR TITLE
HDDS-7624. Reuse MultipartInputStream in BlockInputStream

### PR DESCRIPTION
**Pull request by Cyrill**
## What changes were proposed in this pull request?

The goal of this refactoring is to replace the code to read from a list of streams in BlockInputStream with MultipartInputStream. 

- Instead of storing the list of streams and calculating offsets in-place we delegate to MultipartInputStream
- Now MultipartInputStream has three subclasses, changed the class to match them all

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-7624

## How was this patch tested?

Existing unit tests

### CODE REVIEW

**Comment by Cyrill** [comment]
@adoroszlai Please review

**Comment by kerneltime** [comment]
@Cyrill thank you for the change. I am curious if this is for code reuse/code cleanup or if there are any performance of functional differences desired from this code change.

cc @swamirishi @aswinshakil can you'll please take a look?

**Comment by Cyrill** [comment]
@kerneltime The goal was to reuse the code and make it easier to understand.
I noticed the same code pattern with subtle differences while debugging the ozone client and decided to separate the stream calculation from RPCs.

**Comment by Cyrill** [comment]
@kerneltime @umamaheswararao @jojochuang  Could you please take a look at the changes?

**Comment by Cyrill** [comment]
@kerneltime @umamaheswararao @jojochuang  Could you please take a look at the changes?

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/MultipartCryptoKeyInputStream.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.ozone.client.io;
```
Comment (jojochuang) [inline hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/MultipartCryptoKeyInputStream.java:18]: would it make sense to move over to org.apache.hadoop.hdds.scm.storage package? Similar to MultipartInputStream.

**Review by jojochuang — CHANGES_REQUESTED** [review]
I want to take another look to review the correctness of the code. But it looks like a clever approach. Thanks for working on this.

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/MultipartCryptoKeyInputStream.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * {@link MultipartInputStream} for accessing MPU keys in encrypted buckets.
+ * see also {@link OzoneCryptoInputStream}
+ */
+public class MultipartCryptoKeyInputStream extends MultipartInputStream {
+
+  public MultipartCryptoKeyInputStream(
+      String keyName,
+      List<? extends PartInputStream> inputStreams) {
+    super(" Key: " + keyName, inputStreams);
```
[the remark this answers is not in the document]
Comment (Cyrill) [reply hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/MultipartCryptoKeyInputStream.java:35]: It's used as a prefix in logging and exceptions.

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/MultipartCryptoKeyInputStream.java (hunk 1 above)
Comment (Cyrill) [reply hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/MultipartCryptoKeyInputStream.java:18]: It's used in the client code only, thus added to the client module. The same place the original file was added to (before the refactoring)

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java [hunk 3]
```diff
@@ -279,139 +252,38 @@ protected ChunkInputStream createChunkInputStream(ChunkInfo chunkInfo) {
...
    * not initialized, save the seeked position via blockPosition. Otherwise,
    * update the position in 2 steps:
-   *    1. Updating the chunkIndex to the chunkStream corresponding to the
-   *    seeked position.
-   *    2. Seek the corresponding chunkStream to the adjusted position.
-   *
+   * 1. Updating the chunkIndex to the chunkStream corresponding to the
+   * seeked position.
+   * 2. Seek the corresponding chunkStream to the adjusted position.
+   * <p>
```
Comment (DaveTeng0) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java:268]: Maybe this <p> was added by accident?

**Comment by kerneltime** [comment]
@Cyrill do you plan to continue with this PR? Can you please rebase?

**Pull request closed without merging**